### PR TITLE
PR: Create `UpdateManager` plugin to handle updates for the conda-based installers

### LINKED ISSUE #20831: Migrate existing update mechanism to new conda-based installers
The existing update plugin needs to be updated for the new conda-based installers.
See [this comment](https://github.com/spyder-ide/spyder/issues/20001#issuecomment-1510449004).

1. Migrate the update mechanism as is; do not consider inplace conda updates, just installer download and execution.
2. Modify to trigger download of installer for some releases and trigger inplace conda updates for other releases.
   * What releases should trigger a new installer download and which should trigger a conda update (update inplace)?
   * How should the runtime conda environment be structured/updated?

Should this be split across multiple milestones?

### PR DESCRIPTION
**Pull request by mrclary**
### General Changes

* The existing update mechanism is removed from the `Application` plugin to its own `UpdateManager` plugin. This allows a simplified paradigm where the plugin container can take the role of signal dispatcher between the main `UpdateManagerWidget` and the `UpdateManagerStatus` status bar widget.
* A preference pane was created for the Update Manager that contains "Check for updates at startup" and "Check for stable releases only" preferences. The check at startup preference was removed from the "Application" preference pane.
* The update urls and messaging are essentially unchanged from the recent 5.x PR (#21402, #21423, and #21433).

### Update Flow

* Upon commencing with the install, Spyder will quit and the installation/update process will proceed.
* If the update downloaded a conda-based installer, an interactive graphical uninstaller will launch (Windows) or an interactive command-line uninstaller is executed at a shell terminal (macOS, Linux). For all platforms, the interactive graphical installer is launched upon successful completion of the uninstall process.
* If the update did not require downloading a conda-based installer, a shell terminal will inform the user that Spyder is being updated via conda, and conda's standard output is displayed.
* In all scenarios, Spyder will start/restart upon successful install/update.
* For current versions that are not conda-based installs, only downloading the conda-based installer will result in installing an update. If the user rejects this option, the user is only instructed how to perform the update, but no automatic update will be peformed.

<details><summary>Class Mermaid Map</summary>

```mermaid
classDiagram
  class ProgressDialog { }
  class UpdateManager { }
  class UpdateManagerConfigPage { }
  class UpdateManagerContainer { }
  class UpdateManagerStatus { }
  class UpdateManagerWidget { }
  class WorkerDownloadInstaller { }
  class WorkerUpdate { }
  UpdateManagerConfigPage --* UpdateManager : CONF_WIDGET_CLASS
  UpdateManagerContainer --* UpdateManager : CONTAINER_CLASS
  UpdateManagerStatus --* UpdateManagerContainer : update_manager_status
  ProgressDialog --* UpdateManagerWidget : progress_dialog
  UpdateManagerWidget --* UpdateManagerContainer : update_manager
  WorkerDownloadInstaller --* UpdateManagerWidget : download_worker
  WorkerUpdate --* UpdateManagerWidget : update_worker
```
</details>

### Issue(s) Resolved

Fixes #20831
Supersedes #20900

## Interactive Testing

Interactive testing is most easily performed using the following script. Just copy and save  this somewhere and run it from a Spyder developer environment.
<details><summary>interactive_test.py</summary>

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import logging
from unittest.mock import patch
from spyder.utils.qthelpers import qapplication
from spyder.plugins.updatemanager import workers
from spyder.plugins.updatemanager.widgets import update
from spyder.plugins.updatemanager.widgets.update import UpdateManagerWidget

logging.basicConfig()
_logger = logging.getLogger('spyder.plugins.updatemanager')
_logger.setLevel(10)

parser = argparse.ArgumentParser()
parser.add_argument('-v', '--version', type=str, help="Spyder current version")
parser.add_argument('-s', '--stable-only', action='store_true', default=False,
                    help="Check for stable releases only")
parser.add_argument('-e', '--env', choices=['conda', 'app', 'pip'],
                    default='conda', help='Environment type')
parser.add_argument('-a', '--anaconda', action='store_true', default=False,
                    help="Anaconda pkg installed")
parser.add_argument('-c', '--channel', default='conda-forge',
                    choices=['conda-forge', 'main', 'develop'],
                    help="Spyder conda channel")

cli_opts = parser.parse_args()

if cli_opts.env == 'app':
    is_conda = True
    is_anaconda_pkg = False
    is_conda_based_app = True
    spyder_conda_channel = ('conda-forge',
                            'https://conda.anaconda.org/conda-forge/osx-64')

if cli_opts.env == 'conda':
    is_conda = True
    is_anaconda_pkg = cli_opts.anaconda
    is_conda_based_app = False
    spyder_conda_channel = {
        'conda-forge': ('conda-forge',
                        'https://conda.anaconda.org/conda-forge'),
        'main': ('pkgs/main', 'https://repo.anaconda.com/pkgs/main'),
        'develop': ('pypi', None),  # local install
    }[cli_opts.channel]

if cli_opts.env == 'pip':
    is_conda = False
    is_anaconda_pkg = False
    is_conda_based_app = False
    spyder_conda_channel = (None, None)


def mock_get_conf(self, option, *args, **kwargs):
    if option == 'check_stable_only':
        return cli_opts.stable_only
    else:
        return super(UpdateManagerWidget, self).get_conf(option, *args, **kwargs)


for mod in [workers, update]:
    for attr, new in [
        ('__version__', cli_opts.version),
        ('is_conda_based_app', lambda: is_conda_based_app),
        ('is_anaconda', lambda: is_conda),
    ]:
        patch.object(mod, attr, new=new).start()
patch.object(update, 'is_anaconda_pkg', new=lambda: is_anaconda_pkg).start()
patch.object(workers, 'get_spyder_conda_channel',
             new=lambda: spyder_conda_channel).start()
patch.object(UpdateManagerWidget, 'get_conf', new=mock_get_conf).start()

app = qapplication()
um = UpdateManagerWidget(None)
um.start_check_update()
app.exec()
```
</details>

However, this will only be interactive with the plugin. To test the full interactivity with Spyder, see the interactive instructions in #20900.

### CODE REVIEW

File: spyder/plugins/updatemanager/confpage.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+`main`.
+"""
+
+from qtpy.QtWidgets import QGroupBox, QVBoxLayout
+
+from spyder.config.base import _
+from spyder.api.preferences import PluginConfigPage
+
+
+class UpdateManagerConfigPage(PluginConfigPage):
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/confpage.py:21]: @ccordoba12, thanks for the quick suggestion. Note that
there will be 2 checkboxes: one for check at startup and one for allowing non-stable releases.

I agree that a dedicated configuration page for just 2 checkboxes seems sparse, but I decided this would be best in the interest of making the Update Manager plugin self-contained. However, I can keep the checkboxes in the Application config page if there is agreement to do so.

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/updatemanager/confpage.py:21]: We can discuss about that in our meeting tomorrow, but I'd be in favor of leaving these options in the Application config page.

**Comment by dalthviz** [comment]
Checked this locally and not totally sure if I'm doing things correctly but seems like I'm unable to trigger a major release update 🤔 I was able to trigger a minor release update manually though :+1: 

### Steps followed (extrapolating from the steps at https://github.com/spyder-ide/spyder/pull/20900#issue-1697717335)

1. Install latest 6.0.0a2 (via conda-installer)
1. Update conda-based installation.
   1. Checkout the https://github.com/mrclary/spyder/tree/cbi-updater-plugin  branch locally.
   1. Modify the following files in the Spyder repository:
      - `spyder/__init__.py`. Change the version on line 32 to `(5, 4, 4, "dev0")`.
      - `spyder/plugins/updatemanager/widgets/update.py`. Change the delay interval on line 197 from `60000` ms to `10000` ms. 
      - `spyder/plugins/updatemanager/scripts/install.bat`. Remove `-y` from the conda install command on line 66 (`.bat`) to allow you to cancel a conda update without clobbering your existing install.
   1. Install `superqt` and `qdarkstyle` 3.2 over the `spyder-runtime` env created by the Spyder 6.0.0a2 conda-installer
   1. Run the following command in a terminal/cmd window to update the conda-based Spyder.
      ```cmd
      conda run --live-stream -p "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime" python "E:\Acer\Documentos\Spyder\Spyder otros\mrclary\spyder\install_dev_repos.py" --not-editable
      ```

## Update Flow that partially worked (minor version, check via manual trigger)
1. Start Spyder.
1. The updater should run at Spyder launch (I had to trigger the `Check updates...` action manually otherwise I got the status bar just displaying `Spyder 5.4.4.dev0`).
   1. Testing for a minor update, Spyder will be updated in-place using `conda` mechanisms (worked as described :+1:). 
      - You should be alerted that Spyder 5.4.5 is available and whether you want to proceed to install it. Click "Yes".
      - Spyder will quit.
      - A terminal/cmd window appears and displays the progress of the conda update. After the environment solve, do _not_ confirm the changes: type "n" and hit return. You do not want to proceed with the conda update.
      - After the terminal/cmd window closes, Spyder will launch.

## Update Flow that didn't work (major version - the update available dialog showed me 5.5.0 always)

1. Testing for a major update, a conda-based installer will be downloaded and launched.
      - You should be alerted to Spyder 5.5.0 being available and whether you want to proceed to install it (I had to trigger the `Check updates...` action manually otherwise I got the status bar just displaying `Spyder 5.4.4.dev0` as before). Click "No".
      - Open *Preferences->Update Manager* and uncheck `Check for stable releases only`.
      - Click `Check for updates...` in the Help menu (or clicking over the status bar).
      - You should be alerted that version Spyder 6.0.0a2 is available and whether you want to download and install it (in my tests I always got 5.5.0 🤔 ). Click "Yes". This will initiate downloading the conda-based installer.
      - After the download is complete, a message box will appear asking if you want to proceed with the installation. Click "Yes".
      - Spyder will quit.
      - Spyder's uninstall script will launch interactively in a terminal asking if you wish to uninstall Spyder (macOS or Linux), or a Windows Uninstaller will launch (Windows). Cancel the uninstall operation.
      - After the uninstall terminal/window is closed, the installer should launch. Cancel the installer.

It's possible that I'm missing setting up something (for example I didn't change the updates worker code since seems like the url should point to the GitHub release page which has the alpha release?) so will try to check again

**Comment by mrclary** [comment]
@dalthviz, thanks for looking at this! I think I found the issue. When getting a list of updates from GitHub, the versions are listed in chronologically released order, rather than release order, so `5.5.0` always shows up last:
`['4.2.4', '4.2.5', '5.0.0', '5.0.1', '5.0.2', '5.0.3', '5.0.4', '5.0.5', '5.1.0', '5.1.1', '5.1.2', '5.1.3', '5.1.4', '5.1.5', '5.2.0', '5.2.1', '5.2.2', '5.3.0', '5.3.1', '5.3.2', '5.3.3', '5.4.0', '5.4.1', '5.4.2', '5.4.3', '6.0.0a1', '5.4.4', '5.4.5', '6.0.0a2', '5.5.0']`

I'll fix this and push it up.

**Comment by mrclary** [comment]
@dalthviz, I've pushed up fixes.

**Comment by dalthviz** [comment]
Checking again seems like once a check for updates is started things are working as expected (at least for the most straigh forward flow where you say yes to everything) :+1: Just in case, tested setting the version without `"dev0"` but still seems like the check for updates is not being done automatically even if the `Check for updates on startup` option is enabled 🤔 Couple of screenshots showing how things are looking on Windows with the conda-installers:

## Minor release update:

* Launch Spyder and get the update dialog message (still needs to be lauched manually though 🤔):

<details>

Sin títuloimage

</details>

* Click `Yes` and Spyder closes and a `cmd` showing the update process shows up:

<details>

image

</details>

* Due to testing setup then you get prompt to continue or not with the installation:

<details>

image

</details>

* Spyder is relauched when the `cmd` closes:

<details>

image

</details>

## Major release update (with the option to get only stable release unchecked):

* Launch Spyder and get the update dialog message (still needs to be lauched manually though 🤔):

<details>

Sin títuloimage

</details>

* Click `Yes` and then the download dialog starts:

<details>

image

</details>

* After finishing the install dialog appears:

<details>

image

</details>

* Click `Yes`, Spyder closes and a `cmd` shows as well as the uninstall installer wizard:

<details>

image

</details>

* When the uninstaller closes the installer is launched (regarding this, should we somehow check if Spyder is actually uninstalled and depending on that show the installer or just relaunch Spyder?):

<details>

image

</details>

Will try now to test other actions combinations (like for example cancel the download or select to install the update after closing Spyder) and document anything if I manage to see any weird behaviors or issues

**Comment by mrclary** [comment]
@dalthviz, thanks for following up. I'll investigate why Spyder does not check for updates at startup.

> should we somehow check if Spyder is actually uninstalled and depending on that show the installer or just relaunch Spyder?

I agree that this should be done at some point. We can discuss whether we should defer until later or not, depending on how involved such a check would be.

**Comment by mrclary** [comment]
@dalthviz, I found the issue and pushed up the fix. It was just a typo preventing checking for update at startup.

**Comment by dalthviz** [comment]
Awesome! Indeed, seems like now the automatic check for updates on startup works :+1:

Started checking some other actions combinations and noticed the following when doing a major release update:

* If you click `Yes`, wait for the installer download to finish, then click `After closing` and the click over the status bar you get prompted to download again the installer:

major_update

Maybe we need to save the downloaded installer somehow and skip that step when clicking the status bar if that was already done?

Also, I think I got to see the `cmd` with the output for an update to Spyder 5.5.0 after closing Spyder when the update dialog showed me a major update to 6.0.0a2 although that only happened one time 🤔. I will try to see what could be triggering that or if was just a tests setup error on my side

**Comment by mrclary** [comment]
@dalthviz, if the installer is downloaded, it should not ask to download it again. I'll check to see what the issue is.

**Comment by mrclary** [comment]
@dalthviz, I pushed up the fix. The issue was that verifying the existing download failed due a wrong data type comparison.

To verify the download, I store the download file size read from the url request header and compare this to the actual download file size. This ensures that the download can be verified even if Spyder is restarted or offline. Is there a better way to do this?

**Comment by dalthviz** [comment]
> I pushed up the fix. The issue was that verifying the existing download failed due a wrong data type comparison.

Yep, checked again with the latest changes and now things are working :+1:

> To verify the download, I store the download file size read from the url request header and compare this to the actual download file size. This ensures that the download can be verified even if Spyder is restarted or offline. Is there a better way to do this?

Maybe if we make available from the release the MD5 checksum values for the installers we could use them to check the downloaded file or any previous downloaded installer integrity? Besides that not totally sure 🤔

**Comment by mrclary** [comment]
> Maybe if we make available from the release the MD5 checksum values for the installers we could use them to check the downloaded file or any previous downloaded installer integrity? Besides that not totally sure 🤔

Yeah. I was thinking something similar. Maybe we could explore this type of option in a follow-up PR.

**Comment by dalthviz** [comment]
> Maybe we could explore this type of option in a follow-up PR.

Sound good to me :+1:

Also, I gave this a last manual check and one last thing that maybe could be worthy to give some thought is the handling of the `Check stable versions only` when changed. Maybe it could make sense to restart the update state (status bar message for example) to show again the current version installed when the option is changed? Or even not only that but prompt the users to see if they want to check for updates? Currently what I see is that the previous check result persist even when you change the option. That can lead to some strange interactions where you could be prompt to install a version that doesn't match with the setting. For example, an unstable version was discovered previously and now you check the option to just see if an stable update is available. In that case you would see something like:

update_manager

Besides this behavior seems like everything else is working as expected, at least from the side of the installer on Windows. Will try now to give a check to the code :+1:

**Comment by mrclary** [comment]
@dalthviz, that is a good observation. I'll look into changing the status bar when the preference is changed.

**Comment by mrclary** [comment]
@dalthviz, The statusbar is now updated appropriately when changes to the "Check stable versions only" preference is applied.

File: spyder/config/main.py [hunk 2]
```diff
@@ -78,12 +78,16 @@
...
               'cursor/width': 2,
               'completion/size': (300, 180),
               'report_error/remember_token': False,
               'show_dpi_message': True,
               }),
+            ('update_manager',
+             {
+              'check_updates_on_startup': True,
+              'check_stable_only': True,
+              }),
```
Comment (dalthviz) [inline spyder/config/main.py:90]: I think this requires a `CONF_VERSION` bump?: https://github.com/spyder-ide/spyder/blob/9f1baeb72fae2214f30bccf44aa4aa0c32b7f122/spyder/config/main.py#L660-L670

File: spyder/plugins/updatemanager/confpage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dalthviz) [reply spyder/plugins/updatemanager/confpage.py:17]: Just in case, for the preferences page to not look so empty we could also add some way to show the current Spyder version and also a way to trigger from the preferences a check for updates?

File: spyder/plugins/updatemanager/tests/test_update_manager.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        workers, "get_spyder_conda_channel",
+        return_value=("conda-forge", "https://conda.anaconda.org/conda-forge"))
+
+    um = UpdateManagerWidget(None)
+    um.start_check_update()
+    qtbot.waitUntil(um.update_thread.isFinished)
+
+    _update = um.update_worker.update_available
+    assert _update if version.split('.')[0] == '1' else not _update
+    # assert len(worker.releases) > 1
```
Comment (dalthviz) [inline spyder/plugins/updatemanager/tests/test_update_manager.py:58]: Should this be removed?

**Review by dalthviz — COMMENTED** [review]
Thanks @mrclary ! Checked locally and things are working as expected (including the last change discussed about changing the status when updating the preferences about stable versions checks). Left a couple of comments but other than that this LGTM :+1:

File: spyder/config/main.py (hunk 2 above)
Comment (mrclary) [reply spyder/config/main.py:90]: Good catch @dalthviz. `check_updates_on_startup` is being _removed_ from the `main` section, so I think you are correct and, according to note 2, a major version bump is required. I'll make that change.

File: spyder/plugins/updatemanager/tests/test_update_manager.py (hunk 3 above)
Comment (mrclary) [reply spyder/plugins/updatemanager/tests/test_update_manager.py:58]: Yep. Thanks for catching that.

File: spyder/plugins/updatemanager/confpage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/confpage.py:17]: Those are interesting ideas. Perhaps we should discuss at the next dev meeting.

**Review by dalthviz — APPROVED** [review]
Thanks @mrclary !

**Comment by ccordoba12** [comment]
@mrclary, I was quite advanced in my review when I noticed that, in my opinion, this PR has a big problem regarding its git history, which seriously affects a more careful review and also our future usage of `git blame` for all this code.

The problem is that in your first commit you removed files such as:

* spyder/plugins/application/widgets/install.py
* spyder/plugins/application/widgets/status.py
* spyder/workers/updates.py
* spyder/workers/tests/test_update.py

and then recreated them in your second commit under the new `UpdateManager` plugin. Instead, what you should have done is using `git mv` to move those files and then fix their imports to make them work under the new plugin.

These actions made all those file appear as new, which doesn't allow me to see what you actually changed in them. And as far as I can tell (by copy/pasting a previous and new file in an online diff editor) is quite a bit, but of course not all.

So, my suggestions for you are:

* Create a new branch with an initial commit as the one I mentioned (i.e. using `git mv` and fixing imports).
* Copy/paste the changed files in your second commit here to that branch and do a second commit.
* Cherry-pick every following commit done here to the new branch.

That will show a nice diff for review and also preserve the git history of all moved files for future `git blame`.

Let me know if that's ok for you or you need a hand with it.

**Comment by mrclary** [comment]
@ccordoba12, thanks for looking at this PR.
I understand what you are saying regarding the git history. There was so much refactoring, in addition to moving to the new plugin, that the new files don't really resemble the old ones, so I thought the diffs would be more confusing than helpful.

Nevertheless, I can do as you suggest.
This is what I will do:
* Create a new branch with the recommended preserved history.
* Create a new PR with the new branch for you to inspect.
* If you confirm that the new branch/PR is what you want, then I will just reset the branch of the current PR to the newly created branch, force-push the changes, and discard the newly created PR.

**Comment by mrclary** [comment]
@ccordoba12, actually I just pushed the changes up to this PR. I have a backup branch locally I can reset to if need be.

**Comment by ccordoba12** [comment]
> There was so much refactoring, in addition to moving to the new plugin, that the new files don't really resemble the old ones, so I thought the diffs would be more confusing than helpful.

I understand your reasoning but from what I saw (using a diff editor, as I told you) the diffs of some files at least are readable and important to understand what you exactly changed in this PR.

> Nevertheless, I can do as you suggest.

Yes, please do.

> This is what I will do:

I agree with that and thanks for taking the extra effort of doing it.

> actually I just pushed the changes up to this PR. I have a backup branch locally I can reset to if need be.

I don't understand this message. Are you planning to open a new PR? Or just push directly here with the new history?

**Comment by mrclary** [comment]
> I don't understand this message. Are you planning to open a new PR? Or just push directly here with the new history?

Sorry for the confusion. I just pushed directly here with the new history.

**Comment by ccordoba12** [comment]
@mrclary, this is still a bit of mess. For instance:

* You moved `spyder/plugins/application/widgets/install.py` to `spyder/plugins/updatemanager/widgets/install.py` without modifications. But it seems changes to that file were actually done in `spyder/plugins/updatemanager/widgets/update.py`, which appears as new here.
* The same is true for `spyder/plugins/updatemanager/updates.py` and `spyder/plugins/updatemanager/workers.py` (i.e. the first one moved without modifications and second one where the changes were done).

At this point I don't know if this work could be saved somehow or if it'd be better to start over in a new branch and PR.

**Comment by mrclary** [comment]
> @mrclary, this is still a bit of mess. For instance:
> 
> * You moved `spyder/plugins/application/widgets/install.py` to `spyder/plugins/updatemanager/widgets/install.py` without modifications. But it seems changes to that file were actually done in `spyder/plugins/updatemanager/widgets/update.py`, which appears as new here.
> * The same is true for `spyder/plugins/updatemanager/updates.py` and `spyder/plugins/updatemanager/workers.py` (i.e. the first one moved without modifications and second one where the changes were done).
> 
> At this point I don't know if this work could be saved somehow or if it'd be better to start over in a new branch and PR.

@ccordoba12, you are correct. I did not execute it properly and I overlooked some of those file renames. I apologize for the inconvenience; I'll try one more time.

**Comment by mrclary** [comment]
@ccordoba12, I think this should be good now. I notice for files that both moved _and_ renamed, GitHub diff shows the original as deleted and the new as a new file, i.e. not compared. However, the git history _does_ show the diffs. GitHub diff behaves as expected for files that are only moved. If there are still issues, then perhaps I'm doing something wrong and we should discuss this via zoom. I'll be unavailable tomorrow and Tuesday between 9am-5pm PST, but evenings would be okay.

File: spyder/plugins/updatemanager/workers.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        error_msg = None
+        pypi_url = "https://pypi.org/pypi/spyder/json"
+
+        if is_conda_based_app():
+            url = 'https://api.github.com/repos/spyder-ide/spyder/releases'
+        elif is_anaconda():
+            self.channel, channel_url = get_spyder_conda_channel()
+
+            if channel_url is None or self.channel == "pypi":
+                url = pypi_url
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/workers.py:105]:
```suggestion
            if channel_url is None:
                return
            elif self.channel == "pypi":
                url = pypi_url
```



Why did you change the previous logic here? If `channel_url` is None, we shouldn't recommend to install Spyder from pip. That means that users have it in development mode or that we don't know how they installed it. So, trying to recommend pip could actually break their envs.

File: spyder/plugins/updatemanager/workers.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        except ConnectionError as err:
+            error_msg = CONNECT_ERROR_MSG
+            logger.warning(err, exc_info=err)
+        except HTTPError as err:
+            error_msg = HTTP_ERROR_MSG.format(status_code=page.status_code)
+            logger.warning(err, exc_info=err)
+        except Exception as err:
+            error = traceback.format_exc()
+            formatted_error = (error.replace('\n', '<br>')
+                                    .replace(' ', '&nbsp;'))
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/workers.py:143]:
```suggestion
            formatted_error = (
                error.replace('\n', '<br>')
                .replace(' ', '&nbsp;')
            )
```

File: spyder/plugins/updatemanager/widgets/update.py [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+        if os.name == 'nt':
+            plat, ext = 'Windows', 'exe'
+        if sys.platform == 'darwin':
+            plat, ext = 'macOS', 'pkg'
+        if sys.platform.startswith('linux'):
+            plat, ext = 'Linux', 'sh'
+        mach = platform.machine().lower().replace("amd64", "x86_64")
+        fname = f'Spyder-{self.latest_release}-{plat}-{mach}.{ext}'
+
+        dirname = osp.join(TMPDIR, 'updates', self.latest_release)
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/widgets/update.py:244]:
```suggestion
        dirname = osp.join(get_temp_dir(), 'updates', self.latest_release)
```

File: spyder/plugins/updatemanager/workers.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        except SSLError as err:
+            error_msg = SSL_ERROR_MSG
+            logger.warning(err, exc_info=err)
+        except ConnectionError as err:
+            error_msg = CONNECT_ERROR_MSG
+            logger.warning(err, exc_info=err)
+        except Exception as err:
+            error = traceback.format_exc()
+            formatted_error = (error.replace('\n', '<br>')
+                                    .replace(' ', '&nbsp;'))
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/workers.py:258]:
```suggestion
            formatted_error = (
                error.replace('\n', '<br>')
                .replace(' ', '&nbsp;')
            )
```

File: spyder/plugins/updatemanager/confpage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/updatemanager/confpage.py:17]: I'd prefer to move back these options to the Application config page. We can see later if there's a need to create a new config page, but for now I think it's better to leave them where they are.

File: spyder/plugins/updatemanager/tests/test_update_manager.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    assert not _update if "a" in release and stable_only else _update
+
+
+# ---- Test WorkerDownloadInstaller
+
+@pytest.mark.skipif(not running_in_ci(), reason="Download only in CI")
+def test_download(qtbot, mocker):
+    """
+    Test download spyder installer.
+    Uses UpdateManagerWidget in order to also test QThread.
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/tests/test_update_manager.py:131]: I forgot about that! Perhaps we just skip this test for now and a follow-up PR will use a different repo and url for download testing?

File: spyder/plugins/updatemanager/widgets/update.py [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+
+INSTALL_INFO_MESSAGES = {
+    DOWNLOADING_INSTALLER: _("Downloading Spyder {version}"),
+    DOWNLOAD_FINISHED: _("Finished downloading Spyder {version}"),
+    INSTALLING: _("Installing Spyder {version}"),
+    FINISHED: _("Finished installing Spyder {version}"),
+    PENDING: _("Spyder {version} available to download"),
+    CHECKING: _("Checking for new Spyder version"),
+    CANCELLED: _("Spyder update cancelled"),
+    INSTALL_ON_CLOSE: _("Install Spyder {version} on close")
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/widgets/update.py:56]: This is part of `INSTALL_INFO_MESSAGES` dictionary, that was left over from the the old updater. It is not currently being used and I should just remove it.

File: spyder/plugins/updatemanager/widgets/update.py [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+}
+
+HEADER = _("<h3>Spyder {} is available!</h3><br>")
+URL_I = 'https://docs.spyder-ide.org/current/installation.html'
+TMPDIR = osp.join(gettempdir(), 'spyder')
```
Comment (mrclary) [inline spyder/plugins/updatemanager/widgets/update.py:61]:
```suggestion
TMPDIR = get_temp_dir()
```

File: spyder/plugins/updatemanager/workers.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            error_msg = _(
+                'It was not possible to check for Spyder updates due to the '
+                'following error:'
+                '<br><br>'
+                '<tt>{}</tt>'
+            ).format(formatted_error)
+            logger.warning(err, exc_info=err)
+        finally:
+            self.error = error_msg
+            self.sig_ready.emit()
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/workers.py:154]: Just for my education, when can a signal `emit()` ever raise a `RuntimeError`?

File: spyder/plugins/updatemanager/widgets/update.py [hunk 12]
```diff
@@ -0,0 +1,587 @@
...
+                "install it using Spyder's installer?"
+                "<br><br>"
+                "We <a href='{}'>recommend our own installer</a> "
+                "because it's more stable and makes updating easy. "
+                "This will leave your existing Spyder installation "
+                "untouched."
+            ).format(URL_I + "#standalone-installers")
+            box = confirm_messagebox(
+                self, msg, version=self.latest_release, checkbox=True
+            )
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/widgets/update.py:296]: You want 2 blank lines?

File: spyder/plugins/updatemanager/workers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/updatemanager/workers.py]: It can happen when the object is garbage collected before the signal is emitted.

File: spyder/plugins/updatemanager/widgets/update.py [hunk 13]
```diff
@@ -0,0 +1,603 @@
...
+    box.show()
+    return box
+
+
+def info_messagebox(parent, message, version=None, checkbox=False):
+    box_class = UpdateMessageCheckBox if checkbox else UpdateMessageBox
+    message = HEADER.format(version) + message if version else message
+    box = box_class(icon=QMessageBox.Information, text=message, parent=parent)
+    box.setWindowTitle((_("New Spyder version"))
+    box.setWindowTitle((_("New Spyder version"))
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/widgets/update.py:513]:
```suggestion
    box.setWindowTitle(_("New Spyder version"))
```

This was probably my mistake during review. That's why tests are failing.

Comment (mrclary) [reply spyder/plugins/updatemanager/widgets/update.py:513]: I caught it, but I'm not done with changes yet anyway. Thanks.

File: spyder/plugins/updatemanager/confpage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/plugins/updatemanager/confpage.py:17]: Should the configuration options also be returned to the main section or should we keep the update_manager config section?

**Comment by ccordoba12** [comment]
@mrclary, thanks for your work to address my review promptly! I took a quick look at it and things seem fine to me now.

@dalthviz, could you give this PR a new review to check that @mrclary's latest changes didn't break what was already working? Thanks!

**Comment by dalthviz** [comment]
Giving this a new manual check locally, seems like there is an issue related with the settings (the report error dialog pop-up when unchecking the `Check for stable releases only` option and clicking the preference dialog `Apply` or `Ok` buttons):

```python-traceback
Traceback (most recent call last):
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\plugins\preferences\widgets\configdialog.py", line 185, in button_clicked
    configpage.apply_changes()
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\plugins\preferences\widgets\config_widgets.py", line 119, in apply_changes
    self.apply_callback()
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\plugins\preferences\widgets\config_widgets.py", line 154, in <lambda>
    apply_callback=lambda: self._apply_settings_tabs(
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\plugins\preferences\widgets\config_widgets.py", line 193, in _apply_settings_tabs
    self.apply_settings(options)
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\plugins\application\confpage.py", line 286, in apply_settings
    um = self.plugin.get_plugin(Plugins.UpdateManager, error=False)
  File "C:\Users\dalth\AppData\Local\spyder-6\envs\spyder-runtime\lib\site-packages\spyder\api\plugins\new_api.py", line 444, in get_plugin
    raise SpyderAPIError(
spyder.api.exceptions.SpyderAPIError: Plugin "update_manager" not part of REQUIRES or OPTIONAL requirements!

```

Also, I noticed that there are a couple of `ECHO esta desactivado` lines on the cmd shown when updating a minor version:

image

**Comment by mrclary** [comment]
Thanks @dalthviz. I'll add `UpdateManager` to the `OPTIONAL` list in Application plugin.
Regarding `ECHO is deactivated`, I'm not sure what that is. `@echo off` is used at the top of the script, but that just disables echoing each command that executes 🤷🏼

**Comment by mrclary** [comment]
> Also, I noticed that there are a couple of `ECHO esta desactivado` (`ECHO is deactivated`) lines being shown in the cmd when updating a minor version:

This should be fixed now also.

**Comment by dalthviz** [comment]
Checked again locally and seems like everything is working :tada: so this LGTM :+1:

**Comment by mrclary** [comment]
Thanks @dalthviz and @ccordoba12 for all the feedback!
